text/0517-io-os-reform.md

```diff
@@ -447,7 +457,173 @@ counts, arguments to `main`, and so on).
 ## Revising `Reader` and `Writer`
 [Revising `Reader` and `Writer`]: #revising-reader-and-writer
 
-> To be added in a follow-up PR.
+The `Reader` and `Writer` traits are the backbone of IO, representing
+the ability to (respectively) pull bytes from and push bytes to an IO
+object. The core operations provided by these traits follows a very
+long tradition for blocking IO, but they are still surprisingly subtle
+-- and they need to be revised.
+
+* **Atomicity and data loss**. As discussed
+  [above](#atomicity-and-the-reader-writer-traits), the `Reader` and
+  `Writer` traits currently expose methods that involve multiple
+  actual reads or writes, and data is lost when an error occurs after
+  some (but not all) operations have completed.
+
+  The proposed strategy for `Reader` operations is to return the
+  already-read data together with an error. For writers, the main
+  change is to make `write` only perform a single underlying write
+  (returning the number of bytes written on success), and provide a
+  separate `write_all` method.
+
+* **Parsing/serialization**. The `Reader` and `Writer` traits
+  currently provide a large number of default methods for
+  (de)serialization of various integer types to bytes with a given
+  endianness. Unfortunately, these operations pose atomicity problems
+  as well (e.g., a read could fail after reading two of the bytes
+  needed for a `u32` value).
+
+  Rather than complicate the signatures of these methods, the
+  (de)serialization infrastructure is removed entirely -- in favor of
+  instead eventually introducing a much richer
+  parsing/formatting/(de)serialization framework that works seamlessly
+  with `Reader` and `Writer`.
+
+  Such a framework is out of scope for this RFC, but the
+  endian-sensitive functionality will be provided elsewhere
+  (likely out of tree).
+
+* **The error type**. The traits currently use `IoResult` in their
+  return types, which ties them to `IoError` in particular. Besides
+  being an unnecessary restriction, this type prevents `Reader` and
+  `Writer` (and various adapters built on top of them) from moving to
+  `libcore` -- `IoError` currently requires the `String` type.
+
+  With associated types, there is essentially no downside in making
```

Comment:
Hm. It seems like if we move this into `core`, it couldn't be bounded by `std::error::Error`, right? This could be a problem when interoperating with other pieces of infrastructure that are explicitly bound by `Error`.